File: ariel-backend/app/models/livestream.py
"""
LiveStream Models - Live streaming for educational content
"""
from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime
from enum import Enum


class StreamStatus(str, Enum):
    SCHEDULED = "scheduled"
    LIVE = "live"
    ENDED = "ended"
    CANCELLED = "cancelled"


class StreamCategory(str, Enum):
    STUDY_SESSION = "study_session"
    LECTURE = "lecture"
    Q_AND_A = "q_and_a"
    TUTORIAL = "tutorial"
    EXAM_PREP = "exam_prep"
    DISCUSSION = "discussion"
    OTHER = "other"


class LiveStream(BaseModel):
    """
    Live stream session
    """
    id: Optional[str] = None
    streamer_id: str  # User who is streaming
    title: str
    description: Optional[str] = None
    category: StreamCategory
    subject: Optional[str] = None
    topic: Optional[str] = None

    # Stream details
    status: StreamStatus = StreamStatus.SCHEDULED
    stream_key: Optional[str] = None  # For broadcaster authentication
    stream_url: Optional[str] = None  # RTMP or WebRTC URL
    playback_url: Optional[str] = None  # For viewers
    thumbnail_url: Optional[str] = None

    # Scheduling
    scheduled_start: Optional[datetime] = None
    actual_start: Optional[datetime] = None
    actual_end: Optional[datetime] = None
    duration_minutes: int = 0

    # Engagement
    viewers_count: int = 0
    peak_viewers: int = 0
    likes_count: int = 0
    comments_count: int = 0

    # Settings
    is_public: bool = True
    allow_comments: bool = True
    allow_reactions: bool = True
    save_recording: bool = True

    # Recording
    recording_url: Optional[str] = None

    # Timestamps
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)


class LiveStreamCreate(BaseModel):
    """Request to create/schedule a live stream"""
    title: str
    description: Optional[str] = None
    category: StreamCategory
    subject: Optional[str] = None
    topic: Optional[str] = None
    scheduled_start: Optional[datetime] = None
    is_public: bool = True
    allow_comments: bool = True
    allow_reactions: bool = True
    save_recording: bool = True


class LiveStreamUpdate(BaseModel):
    """Update stream details"""
    title: Optional[str] = None
    description: Optional[str] = None
    category: Optional[StreamCategory] = None
    subject: Optional[str] = None
    topic: Optional[str] = None
    scheduled_start: Optional[datetime] = None
    thumbnail_url: Optional[str] = None
    is_public: Optional[bool] = None
    allow_comments: Optional[bool] = None
    allow_reactions: Optional[bool] = None


class LiveStreamWithStreamer(BaseModel):
    """Stream with streamer info for display"""
    id: str
    streamer_id: str
    streamer_username: str
    streamer_profile_picture: Optional[str]
    streamer_verified: bool

    title: str
    description: Optional[str]
    category: StreamCategory
    subject: Optional[str]
    topic: Optional[str]

    status: StreamStatus
    playback_url: Optional[str]
    thumbnail_url: Optional[str]

    scheduled_start: Optional[datetime]
    actual_start: Optional[datetime]
    duration_minutes: int

    viewers_count: int
    peak_viewers: int
    likes_count: int
    comments_count: int

    is_public: bool
    allow_comments: bool
    allow_reactions: bool

    created_at: datetime

    # Viewer-specific
    is_liked_by_current_user: bool = False
    is_following_streamer: bool = False


class StreamComment(BaseModel):
    """Live stream comment/chat message"""
    id: Optional[str] = None
    stream_id: str
    user_id: str
    username: str
    profile_picture: Optional[str] = None
    message: str
    timestamp: datetime = Field(default_factory=datetime.utcnow)

    # Engagement
    likes: int = 0
    is_pinned: bool = False
    is_deleted: bool = False


class StreamReaction(BaseModel):
    """Real-time reaction (like, heart, clap, etc.)"""
    id: Optional[str] = None
    stream_id: str
    user_id: str
    reaction_type: str  # "like", "heart", "clap", "fire", etc.
    timestamp: datetime = Field(default_factory=datetime.utcnow)

File: ariel-mobile/src/shared/types/livestream.ts
export enum StreamStatus {
  SCHEDULED = 'scheduled',
  LIVE = 'live',
  ENDED = 'ended',
  CANCELLED = 'cancelled',
}

export enum StreamCategory {
  STUDY_SESSION = 'study_session',
  LECTURE = 'lecture',
  Q_AND_A = 'q_and_a',
  TUTORIAL = 'tutorial',
  EXAM_PREP = 'exam_prep',
  DISCUSSION = 'discussion',
  OTHER = 'other',
}

export interface LiveStream {
  id: string | null;
  streamer_id: string;
  title: string;
  description: string | null;
  category: StreamCategory;
  subject: string | null;
  topic: string | null;
  status: StreamStatus;
  stream_key: string | null;
  stream_url: string | null;
  playback_url: string | null;
  thumbnail_url: string | null;
  scheduled_start: string | null;
  actual_start: string | null;
  actual_end: string | null;
  duration_minutes: number;
  viewers_count: number;
  peak_viewers: number;
  likes_count: number;
  comments_count: number;
  is_public: boolean;
  allow_comments: boolean;
  allow_reactions: boolean;
  save_recording: boolean;
  recording_url: string | null;
  created_at: string;
  updated_at: string;
}

export interface LiveStreamWithStreamer {
  id: string;
  streamer_id: string;
  streamer_username: string;
  streamer_profile_picture: string | null;
  streamer_verified: boolean;
  title: string;
  description: string | null;
  category: StreamCategory;
  subject: string | null;
  topic: string | null;
  status: StreamStatus;
  playback_url: string | null;
  thumbnail_url: string | null;
  scheduled_start: string | null;
  actual_start: string | null;
  duration_minutes: number;
  viewers_count: number;
  peak_viewers: number;
  likes_count: number;
  comments_count: number;
  is_public: boolean;
  allow_comments: boolean;
  allow_reactions: boolean;
  created_at: string;
  is_liked_by_current_user: boolean;
  is_following_streamer: boolean;
}

export interface StreamComment {
  id: string | null;
  stream_id: string;
  user_id: string;
  username: string;
  profile_picture: string | null;
  message: string;
  timestamp: string;
  likes: number;
  is_pinned: boolean;
  is_deleted: boolean;
}

export interface StreamViewer {
  user_id: string;
  username: string;
  profile_picture: string | null;
  joined_at: string;
}

// WebSocket message types for live streams
export type LiveWsMessageType =
  | 'viewer_joined'
  | 'chat_message'
  | 'reaction'
  | 'stream_ended'
  | 'viewer_count_update'
  | 'comment'
  | 'viewer_count';

export interface LiveWsViewerJoined {
  type: 'viewer_joined';
  user_id: string;
  username: string;
}

export interface LiveWsChatMessage {
  type: 'chat_message' | 'comment';
  stream_id: string;
  user_id: string;
  username: string;
  profile_picture: string | null;
  message: string;
  timestamp: string;
}

export interface LiveWsReaction {
  type: 'reaction';
  reaction_type: string;
  user_id: string;
}

export interface LiveWsStreamEnded {
  type: 'stream_ended';
}

export interface LiveWsViewerCountUpdate {
  type: 'viewer_count_update' | 'viewer_count';
  count: number;
}

export type LiveWsMessage =
  | LiveWsViewerJoined
  | LiveWsChatMessage
  | LiveWsReaction
  | LiveWsStreamEnded
  | LiveWsViewerCountUpdate;
